infra/Pos/Slotting/MemState/Class.hs

```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
 import           Universum
 
 import           Pos.Core.Types      (EpochIndex, Timestamp)
-import           Pos.Slotting.Types  (SlottingData)
+import           Pos.Slotting.Types  (EpochSlottingData)
+
 
 -- | 'MonadSlotsData' provides access to data necessary for slotting to work.
+-- Both _current_ and _next_ epoch @SlottingData@ need to be present since it's impossible
+-- to create @SlottingData@ without them.
 class Monad m => MonadSlotsData m where
 
-    getSystemStart :: m Timestamp
+    getSystemStartM :: m Timestamp
+
+    getAllEpochIndexM :: m [EpochIndex]
+
+    getCurrentEpochIndexM :: m EpochIndex
+
+    getNextEpochIndexM :: m EpochIndex
 
-    getSlottingData :: m SlottingData
+    getCurrentEpochSlottingDataM :: m EpochSlottingData
 
-    waitPenultEpochEquals :: EpochIndex -> m ()
+    getNextEpochSlottingDataM :: m EpochSlottingData
 
-    putSlottingData :: SlottingData -> m ()
+    getEpochSlottingDataM :: EpochIndex -> m (Maybe EpochSlottingData)
+
+    putEpochSlottingDataM :: EpochIndex -> EpochSlottingData -> m ()
+
+    waitCurrentEpochEqualsM :: EpochIndex -> m ()
 
 
 instance {-# OVERLAPPABLE #-}
     (MonadSlotsData m, MonadTrans t, Monad (t m)) =>
         MonadSlotsData (t m) where
-    getSystemStart = lift getSystemStart
-    getSlottingData = lift getSlottingData
-    waitPenultEpochEquals = lift . waitPenultEpochEquals
-    putSlottingData = lift . putSlottingData
+
+    getSystemStartM = lift getSystemStartM
+
+    getAllEpochIndexM = lift getAllEpochIndexM
+
+    getCurrentEpochIndexM = lift getCurrentEpochIndexM
+
+    getNextEpochIndexM = lift getNextEpochIndexM
+
+    getCurrentEpochSlottingDataM = lift getCurrentEpochSlottingDataM
+
+    getNextEpochSlottingDataM = lift getCurrentEpochSlottingDataM
+
+    getEpochSlottingDataM = lift . getEpochSlottingDataM
+
+    putEpochSlottingDataM = (lift .) . putEpochSlottingDataM
```

Comment:
Btw, there is `...` in `Universum` which is a var-arg generalization of `.`. So you can right `lift ... f` regadless of the number of arguments `f` has.